service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java

```diff
@@ -164,6 +179,103 @@ private void processCredentialsMessageGetOperation(final Message<JsonObject> reg
+        }
+        updateCredentials(tenantId, payload, result -> reply(regMsg, result)); 
+    }
+    
+    private void processCredentialsMessageRemoveOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        final String deviceId = payload.getString(FIELD_DEVICE_ID);
+        if (deviceId == null) {
+            log.debug("credentials remove request did not contain device-id in payload - not supported");
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        
+        final String type = payload.getString(FIELD_TYPE);
+        if (type == null) {
+            log.debug("credentials remove request did not contain device-id in payload - not supported");
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+
+        final String authId = payload.getString(FIELD_AUTH_ID);
+        
+        removeCredentials(tenantId, deviceId, type, authId, result -> reply(regMsg, result));
+    }
+
+    boolean isValidCredentialsObject(JsonObject credentials) {
+        return containsStringValueForField(credentials, RequestResponseApiConstants.FIELD_DEVICE_ID)
+                && containsStringValueForField(credentials, FIELD_TYPE)
+                && containsStringValueForField(credentials, FIELD_AUTH_ID)
+                && containsValidSecretValue(credentials);
+    }
+
+    boolean containsValidSecretValue(JsonObject credentials) {
+        JsonArray secrets = credentials.getJsonArray(FIELD_SECRETS);
+        
+        if(secrets == null) {
+            log.debug("credentials request did not contain {} in payload - not supported", FIELD_SECRETS);
+            return false;
+        }
+
+        for (int i = 0; i < secrets.size(); i++) {
+            JsonObject currentSecret = secrets.getJsonObject(i);
+            if (!containsValidTimestampIfPresentForField(currentSecret, FIELD_SECRETS_NOT_BEFORE)
+                    || !containsValidTimestampIfPresentForField(currentSecret, FIELD_SECRETS_NOT_AFTER)) {
+                log.debug("credentials request did contain invalid timestamp values in payload");
+                return false;
+            }
+        }
+        return true;
+    }
+
+    boolean containsStringValueForField(JsonObject payload, String field) {
+        String value = payload.getString(field);
+
+        if (StringUtils.isEmpty(value)) {
+            log.debug("credentials request did not contain {} in payload - not supported", field);
+            return false;
+        }
+
+        return true;
+    }
+
+    boolean containsValidTimestampIfPresentForField(JsonObject payload, String field) {
+        String timeStamp = payload.getString(field);
+        if (timeStamp != null) {
+            return isValidTimestampForField(payload, field);
+        }
+        else {
+            return true;
+        }
+    }
+
+    boolean isValidTimestampForField(JsonObject payload, String field) {
```

Comment:
why not private?

Reply:
finals

Reply:
Package private was chosen, so that it's possible to unit test the method. The parameters i will make final.